// client/src/utils/countries.js
export const COUNTRIES = {
    USA: { name: 'USA', color: '#B22234', flag: 'https://flagcdn.com/w320/us.png' },
    IND: { name: 'India', color: '#FF9933', flag: 'https://flagcdn.com/w320/in.png' },
    CHN: { name: 'China', color: '#DE2910', flag: 'https://flagcdn.com/w320/cn.png' },
    GBR: { name: 'UK', color: '#012169', flag: 'https://flagcdn.com/w320/gb.png' },
    JPN: { name: 'Japan', color: '#BC002D', flag: 'https://flagcdn.com/w320/jp.png' },
    GER: { name: 'Germany', color: '#000000', flag: 'https://flagcdn.com/w320/de.png' },
    FRA: { name: 'France', color: '#0055A4', flag: 'https://flagcdn.com/w320/fr.png' },
    BRA: { name: 'Brazil', color: '#009739', flag: 'https://flagcdn.com/w320/br.png' },
    ITA: { name: 'Italy', color: '#009246', flag: 'https://flagcdn.com/w320/it.png' },
    CAN: { name: 'Canada', color: '#C52032', flag: 'https://flagcdn.com/w320/ca.png' },
    ESP: { name: 'Spain', color: '#AA151B', flag: 'https://flagcdn.com/w320/es.png' },
    AUS: { name: 'Australia', color: '#012169', flag: 'https://flagcdn.com/w320/au.png' },
};
